Fix: Ensure hf_hub_download arguments are used when loading locally

henryzhengr opened the pull request:
This PR solves the issue that:
When loading a PEFT model locally, the provided arguments for hf_hub_download (e.g., local_dir and cache_dir) are not being passed correctly. As a result, custom paths specified by the user are ignored, leading to potential issues with file loading.

BenjaminBossan commented:
Thanks for the PR. Could you please provide an example where the current code would fail? Also, do you know which parameters specifically from `hf_hub_download_kwargs` are required to be passed to `hf_hub_download`?

henryzhengr commented:
When loading the model using the following code with the environment variable HF_HUB_OFFLINE=1, the current implementation defaults to the Hugging Face cache directory instead of the user-provided cache_dir. This causes issues when trying to load models from a custom directory.

```
checkpoint_dir = "checkpoints/Cosmos-1.0-Guardrail/aegis"
base_model_id = "meta-llama/LlamaGuard-7b"
aegis_adapter = "nvidia/Aegis-AI-Content-Safety-LlamaGuard-Defensive-1.0"

base_model = AutoModelForCausalLM.from_pretrained(base_model_id, cache_dir=checkpoint_dir)
model = PeftModel.from_pretrained(base_model, aegis_adapter, cache_dir=checkpoint_dir)
```

BenjaminBossan commented:
Thanks for providing the example. We should create a unit test based on that to ensure that the fix works and keeps on working. However, I'm not quite sure I can reproduce quite yet. This is what I came up with:

```python
import os
from peft import PeftModel
from transformers import AutoModelForCausalLM

class Test2373:
    def test_it(self):
        tmp_path = "/tmp/peft/2373"
        model_id = "hf-internal-testing/tiny-random-OPTForCausalLM"
        lora_id = "peft-internal-testing/tiny-OPTForCausalLM-lora"

        # first load the LoRA weights from the hub, then create a local checkpoint
        model = AutoModelForCausalLM.from_pretrained(model_id, cache_dir=tmp_path)
        model = PeftModel.from_pretrained(model, lora_id)
        model.save_pretrained(tmp_path)
        del model

        # load the local checkpoint from non-default cache dir
        os.environ["HF_HUB_OFFLINE"] = "1"
        model = AutoModelForCausalLM.from_pretrained(model_id, cache_dir=tmp_path)
        model = PeftModel.from_pretrained(model, lora_id, cache_dir=tmp_path)
```

Any idea what is different here?

Also, for your initial example, why can't you change the last line to:

```diff
- model = PeftModel.from_pretrained(base_model, aegis_adapter, cache_dir=checkpoint_dir)
+ model = PeftModel.from_pretrained(base_model, checkpoint_dir)
```

henryzhengr commented:
If the `PeftModel` is saved using `model.save_pretrained(tmp_path)`, the model files will be stored directly in `tmp_path`. However, when saving the model using methods like:  

- `huggingface_hub.snapshot_download`,  
- `PeftModel.from_pretrained(..., cache_dir=checkpoint_dir)`,  
- or the Hugging Face CLI API,  

the files will be saved under a subdirectory following the format:  

```python
os.path.join(tmp_path, "models--peft-internal-testing--tiny-OPTForCausalLM-lora")
```

Thus, in this case, passing the `cache_dir` argument maybe helpful.

Moreover, setting the environment variable `HF_HUB_OFFLINE=1` **after** package initialization might not work as expected unless the value of `huggingface_hub.constants.HF_HUB_OFFLINE` is explicitly updated. I think to ensure proper offline behavior, it may be necessary to update `huggingface_hub.constants.HF_HUB_OFFLINE` in the code after modifying the environment variable or set it **before** importing Hugging Face-related packages?

BenjaminBossan commented:
Thanks for the additional info. What you mentioned about setting `HF_HUB_OFFLINE` is true. I still have trouble coming up with a unit test, which we really want to add for this PR. At first I tried [monkey patching](https://github.com/huggingface/peft/blob/6d033600e731cca368e69349e85a924b5cd72392/tests/test_initialization.py#L1572) `huggingface_hub.constants.HF_HUB_OFFLINE`, but the test passed even without your fix, so that's not enough. Then I came up with the following test which needs to run in 2 steps:

```python
class Test2373:
    model_id = "hf-internal-testing/tiny-random-OPTForCausalLM"
    lora_id = "peft-internal-testing/tiny-OPTForCausalLM-lora"
    tmp_path = "/tmp/peft/2373"

    def test_step0(self):
        snapshot_download(self.model_id, cache_dir=self.tmp_path)
        snapshot_download(self.lora_id, cache_dir=self.tmp_path)

    def test_step1(self):
        cache_dir = "/tmp/peft/2373"
        model = AutoModelForCausalLM.from_pretrained(self.model_id, cache_dir=self.tmp_path)
        model = PeftModel.from_pretrained(model, self.lora_id, cache_dir=self.tmp_path)
```

The idea is to first run step 0 without offline mode to snapshot download the checkpoints to the cache dir, then stop 1 with offline mode to load from the cache dir:

```sh
HF_HUB_OFFLINE=0 pytest foo.py -k step0
HF_HUB_OFFLINE=1 pytest foo.py -k step1
```

However, even this test passed for me. Do you have a suggestion how else to test this?

henryzhengr commented:
I suspect that maybe these files already exist in your default cache directory?

BenjaminBossan commented:
> I suspect that maybe these files already exist in your default cache directory?

Yes, that was it. So for the unit test to work reliably, we would need to ensure that the LoRA checkpoint does not exist in the default cache location. I tried to check if `huggingface_hub` has a function to delete checkpoints. However, I only found a [CLI command](https://huggingface.co/docs/huggingface_hub/v0.28.1/en/guides/manage-cache#clean-cache-from-the-terminal). @Wauplin is there a way to programmatically delete a checkpoint without having to re-implement the CLI command?

Wauplin commented on src/peft/utils/save_and_load.py:
```diff
@@ -506,13 +506,13 @@ def get_hub_filename(use_safetensors=True):
         hub_filename = get_hub_filename(use_safetensors=True)
         try:
-            filename = hf_hub_download(model_id, hub_filename, local_files_only=True)
+            filename = hf_hub_download(model_id, hub_filename, local_files_only=True, **hf_hub_download_kwargs)
```
I suspect this line will fail if `hf_hub_download_kwargs` contains a key `local_files_only` which is the case if someone does `AutoModelForCausalLM.from_pretrained(model_id, local_files_only=True)`.

BenjaminBossan commented:
> What I would advice you to do is to define a root-level fixture that sets the default cache to a tmp directory for all tests. This way your default local cache will never interfere with the test suite.

Thanks for the tip. I'm not so concerned about my local cache, but rather that on CI, depending on the order of the tests, another test has already loaded the mentioned adapter. In those circumstances, a root level global cache would not help, but the same idea should work for a test specific fixture, right?

> I suspect this line will fail if hf_hub_download_kwargs contains a key local_files_only which is the case if someone does AutoModelForCausalLM.from_pretrained(model_id, local_files_only=True).

Yes, I was thinking about just picking the required kwargs from the dict instead of passing it directly.

Wauplin commented:
> but the same idea should work for a test specific fixture, right?

Yes indeed, better to scope to "function".

BenjaminBossan commented:
Thanks for the help @Wauplin. With this, I could create a test that will fail with current `main` and is fixed by this PR:

```python
class Test2373:
    model_id = "hf-internal-testing/tiny-random-OPTForCausalLM"
    lora_id = "peft-internal-testing/tiny-OPTForCausalLM-lora"

    @pytest.fixture
    def changed_default_cache_dir(self, tmp_path, monkeypatch):
        # ensure that this test does not interact with other tests that may use the HF cache
        monkeypatch.setattr("huggingface_hub.constants.HF_HOME", tmp_path)
        monkeypatch.setattr("huggingface_hub.constants.HF_HUB_CACHE", tmp_path / "hub")
        monkeypatch.setattr("huggingface_hub.constants.HF_TOKEN_PATH", tmp_path / "token")

    @contextmanager
    def hub_offline_ctx(self):
        with patch("huggingface_hub.constants.HF_HUB_OFFLINE", True):
            reset_sessions()
            yield
        reset_sessions()

    def load_checkpoints(self, cache_dir):
        snapshot_download(self.model_id, cache_dir=cache_dir)
        snapshot_download(self.lora_id, cache_dir=cache_dir)

    def test_load_checkpoint_offline_non_default_cache_dir(self, changed_default_cache_dir, tmp_path):
        # See #2373 for context
        self.load_checkpoints(tmp_path)
        with self.hub_offline_ctx():
            model = AutoModelForCausalLM.from_pretrained(self.model_id, cache_dir=tmp_path)
            model = PeftModel.from_pretrained(model, self.lora_id, cache_dir=tmp_path)
```

@henryzhengr Could you please add the unit test to this PR? Ideally, you can add it this existing test:

https://github.com/huggingface/peft/blob/6d033600e731cca368e69349e85a924b5cd72392/tests/test_initialization.py#L1565

Note that the `hub_offline_ctx` method is already defined there, so no need to add it twice.

henryzhengr commented:
Thanks, @BenjaminBossan and @Wauplin, for the guidance! I’ve added the unit test as suggested.  

Additionally, I updated `test_load_from_hub_then_offline_model` to use the same dummy model for improved code cleanliness and consistency. Let me know if this change might introduce any issues.  

> Yes, I was thinking about just picking the required kwargs from the dict instead of passing it directly. 

Would it be better to pass only `cache_dir` if it exists? Do you think any other parameters are essential for proper offline loading?

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/peft/pr_2373). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

BenjaminBossan reviewed (changes requested):
Thanks for adding the test. The update to the previous one looks fine. Could you please run `make style` to satisfy the linter?

> I suspect this line will fail if hf_hub_download_kwargs contains a key local_files_only which is the case if someone does AutoModelForCausalLM.from_pretrained(model_id, local_files_only=True).

> Would it be better to pass only cache_dir if it exists? Do you think any other parameters are essential for proper offline loading?

I think the easiest way is to pop `local_files_only` from `hf_hub_download_kwargs` and pass the remaining kwargs.

BenjaminBossan reviewed (approved):
Thanks for the fix, LGTM.
